File: src/main/java/com/cartrecipes/service/cart/CartService.java
package com.cartrecipes.service.cart;

import com.cartrecipes.exception.DuplicateRecipeException;
import com.cartrecipes.exception.NotFoundException;
import com.cartrecipes.model.Cart;
import com.cartrecipes.model.Recipe;
import com.cartrecipes.service.recipe.RecipeService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.Set;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
public class CartService {
    private final CartRepository cartRepository;

    private final RecipeService recipeService;

    @Transactional(readOnly = true)
    public Cart findWithRecipesAndProductsById(Long id) {
        return cartRepository.findWithRecipesAndProductsById(id)
                .orElseThrow(() -> new NotFoundException("Cart not found with id: " + id));
    }

    public Cart addRecipes(Long cartId, Set<Long> recipeIds) {
        var recipes = recipeService.findByIds(recipeIds);

        if (recipeIds.size() != recipes.size()) {
            var foundRecipeIds = recipes.stream()
                    .map(Recipe::getId)
                    .collect(Collectors.toSet());
            var missingRecipeIds = new ArrayList<>(recipeIds);
            missingRecipeIds.removeIf(foundRecipeIds::contains);
            throw new NotFoundException("One or more recipes not found: " + missingRecipeIds);
        }

        var cart = cartRepository.findWithRecipesAndProductsById(cartId)
                .orElseThrow(() -> new NotFoundException("Cart not found with id: " + cartId));

        var duplicateRecipeIds = recipes.stream()
                .filter(recipe -> !cart.getRecipes().add(recipe))
                .map(Recipe::getId)
                .collect(Collectors.toSet());

        if (!duplicateRecipeIds.isEmpty()) {
            throw new DuplicateRecipeException(duplicateRecipeIds);
        }

        cart.setTotalInCents(cart.getTotalInCents() + recipes.stream()
                .mapToInt(Recipe::getPriceInCents)
                .sum());

        return cartRepository.save(cart);
    }

    public Cart removeRecipe(Long cartId, Long recipeId) {
        var cart = cartRepository.findWithRecipesAndProductsById(cartId)
                .orElseThrow(() -> new NotFoundException("Cart not found with id: " + cartId));

        var recipe = recipeService.findByIds(Set.of(recipeId)).stream()
                .findFirst()
                .orElseThrow(() -> new NotFoundException("Recipe not found with id: " + recipeId));

        if (!cart.getRecipes().remove(recipe)) {
            throw new NotFoundException("Recipe not found in the cart");
        }

        cart.setTotalInCents(cart.getTotalInCents() - recipe.getPriceInCents());

        return cartRepository.save(cart);
    }
}

File: src/main/kotlin/com/cartrecipes/service/cart/CartService.kt
package com.cartrecipes.service.cart

import com.cartrecipes.exception.DuplicateRecipeException
import com.cartrecipes.exception.NotFoundException
import com.cartrecipes.model.Cart
import com.cartrecipes.service.recipe.RecipeService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class CartService internal constructor(
    private val cartRepository: CartRepository,
    private val recipeService: RecipeService
) {

    fun addRecipes(cartId: Long, recipeIds: Collection<Long>): Cart {
        val recipes = recipeService.findByIds(recipeIds)

        if (recipes.size != recipeIds.size) {
            val foundRecipeIds = recipes.map { it.id }.toSet()
            val missingRecipeIds = recipeIds.filterNot { it in foundRecipeIds }
            throw NotFoundException("One or more recipes not found: $missingRecipeIds")
        }

        val cart = cartRepository.findWithRecipesAndProductsById(cartId) ?: throw NotFoundException("Cart not found with id: $cartId")

        val duplicateRecipeIds = recipes
            .filterNot { cart.recipes.add(it) }
            .map { it.id }

        if (duplicateRecipeIds.isNotEmpty()) {
            throw DuplicateRecipeException(duplicateRecipeIds)
        }

        cart.totalInCents += recipes.sumOf { it.priceInCents }

        return cartRepository.save(cart)
    }

    fun removeRecipe(cartId: Long, recipeId: Long): Cart {
        val cart = cartRepository.findWithRecipesAndProductsById(cartId) ?: throw NotFoundException("Cart not found with id: $cartId")
        val recipe = recipeService.findById(recipeId)

        if (!cart.recipes.remove(recipe)) {
            throw NotFoundException("Recipe not found in the cart")
        }

        cart.totalInCents -= recipe.priceInCents

        return cartRepository.save(cart)
    }

    fun findWithRecipesAndProductsById(cartId: Long): Cart {
        return cartRepository.findWithRecipesAndProductsById(cartId)
            ?: throw NotFoundException("Cart not found with id: $cartId")
    }
}